# Dectecting Spam Email.


In [97]:
import pandas as pd
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix
from sklearn.naive_bayes import MultinomialNB


In [98]:
data=pd.read_csv('Data/emails.csv')

In [99]:
print("Information about Data:\n")
print(data.info())

Information about Data:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5728 entries, 0 to 5727
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    5728 non-null   object
 1   spam    5728 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 89.6+ KB
None


In [100]:
data.head()

text  spam
0  Subject: naturally irresistible your corporate...     1
1  Subject: the stock trading gunslinger  fanny i...     1
2  Subject: unbelievable new homes made easy  im ...     1
3  Subject: 4 color printing special  request add...     1
4  Subject: do not have money , get software cds ...     1

In [101]:
data.spam.value_counts()

0    4360
1    1368
Name: spam, dtype: int64

Out of **5728** emails **4360** are not spam and **1368** are **spam**

###### Splitting data 

In [102]:
train_X,val_X,train_y,val_y=train_test_split(data.text,data.spam)

In [103]:
print(train_X.shape,val_X.shape,train_y.shape,val_y.shape)

(4296,) (1432,) (4296,) (1432,)


#### Using CountVectorizer

In [104]:
from sklearn.feature_extraction.text import CountVectorizer
v=CountVectorizer()
train_X_count=v.fit_transform(train_X.values)

In [105]:
train_X_count.toarray()[:5]

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [106]:
from sklearn.naive_bayes import MultinomialNB
nb=MultinomialNB(alpha=1.0, fit_prior=True, class_prior=None)
nb.fit(train_X_count,train_y)

MultinomialNB()

## Applying on test data

In [107]:
mail=["How are you my freind",
      "You have won  $20000 click to accept the money"]
mail_count=v.transform(mail)
nb.predict(mail_count)

array([0, 1])

In [108]:
val_X_count=v.transform(val_X)
a=nb.predict(val_X_count)
nb.score(val_X_count,val_y)

0.99162011173184361

### Using python pipeline

In [109]:
from sklearn.pipeline import Pipeline
clf=Pipeline([
    ("vectorizer",CountVectorizer()),
    ("nb",MultinomialNB())
])


In [110]:
clf.fit(train_X,train_y)

Pipeline(steps=[('vectorizer', CountVectorizer()), ('nb', MultinomialNB())])

In [111]:
preds=clf.predict(val_X)

In [117]:
print('CONFUSION MATRIX:')
print(confusion_matrix(preds,val_y),"\n")
print("Score:{}".format(clf.score(val_X,val_y)))

CONFUSION MATRIX:
[[1061    4]
 [   8  359]] 

Score:0.9916201117318436


**Using gaussianNB**


In [124]:
from sklearn.feature_extraction.text import CountVectorizer
v=CountVectorizer()
train_X_count=v.fit_transform(train_X.values)

In [128]:
train_X_count=train_X_count.todense()

In [122]:
clfg.fit(train_X,train_y)

TypeError: A sparse matrix was passed, but dense data is required. Use X.toarray() to convert to a dense numpy array.